finat/point_set.py

```diff
@@ -69,12 +69,28 @@ def expression(self):
 
 
 class UnknownPointSingleton(AbstractPointSet):
-    """A single unknown point."""
+    """A point set representing a single point with unknown location but known
+    `gem.Variable` expression.
+
+    Stored points are `numpy.nan`s to indicate that they are not known.
+
+    In the future this can be easily replaced with an equivalent for a whole
+    point set representing multiple points though storing the full array of
+    `numpy.nans` may become undesirable and require interface changes."""
 
     def __init__(self, point_expr):
+        """Build a PointSingleton from a gem expression for a single point.
+
+        :arg point_expr: A `gem.Variable` expression representing a single
+            point of shape (D,) where D is the dimension of the point.
+            Since this represents a single point and is a `gem.Variable` it
+            should have no free indices."""
         assert isinstance(point_expr, gem.Variable)
+        # Check we only have 1 point
+        assert point_expr.free_indices == ()
+        assert len(point_expr.shape) == 1
         self.expression = point_expr
-        self.points = numpy.full((1, 1), numpy.nan)
+        self.points = numpy.full(point_expr.shape, numpy.nan).reshape(1, -1)
```

Comment:
If anyone can think of a nicer way of doing this without having to actually allocate memory I'm all ears. As I say in the comment, extending this to a full `PointSet` of unknown points should be pretty easy but it would be daft to be carrying around an array of `NaN`s

Reply:
Perhaps the `AbstractPointSet` needs to be adjusted to not require a cached `numpy.array` as the `points` property

Reply:
When do you want to be able to realise that the `points` property is invalid? I would `raise ValueError("Unknown points do not have a point array")` in the property.

Reply:
Yes I tried that, the problem is that `points` is used all over the codebase to check for things like length matching.

Reply:
If you need something that advertises the "meta-data" part of the numpy array just make a mock object.

Reply:
good idea